plugins/inputs/t128_tank/reader.go

```diff
@@ -154,10 +154,62 @@ func (r *Reader) Run(mainCtx context.Context) {
 				r.lastSavedIndex = lastIndex.value
 			}
 			nextSaveCheck = time.After(2 * time.Second)
+		case err := <-r.readDone:
+			var errBoundaryFault *boundaryFault
+			if err != nil && errors.As(err, &errBoundaryFault) {
+				r.log.Debugf("detected boundary fault, restarting %s tank read from index %d", r.topic, errBoundaryFault.nextAvailableIndex.value)
+				go r.read(mainCtx, readCtx, errBoundaryFault.nextAvailableIndex)
```

Comment:
I missed it earlier. `read` shouldn't take two contexts. Since the read context is created from the main context (`readCtx, readCtxCancel := context.WithCancel(mainCtx)`), canceling the main context will also cancel the read context. So you don't need to pass both.

Reply:
oh actually missed this just now!! didn't realize this while pushing! thanks for catching.

Reply:
P.S. working on unit tests!